### PR TITLE
Commit dialog, FileStatusList: Fix up/down selects unexpected file

### LINKED ISSUE #6212: The method "RestoreSelectedFiles" should always update "SelectedIndex" of both staged and unstaged file list
## Current behaviour

Unstaging a file in the commit window makes that file *appear* selected in "unstaged file list" area. But the truth is that the file is only highlighted, not *selected*. The actual selected file is the first one in the list.

## Expected behaviour

The unstaged file should both be the highlighted and the selected file.

## Steps to reproduce

1. Make changes in several files (like at least 4) in a repository.
2. Open gitextensions commit window.
3. Stage the last file in the list (Either double click on the file name or highlight the file and press <kbd>S</kbd>.
4. Unstage that file by either double clicking or pressing <kbd>U</kbd>.
5. Press <kbd>&uarr;</kbd> and see that nothing happens (because the actual selected file is the first one).
6. Press <kbd>&darr;</kbd> and see that the second file gets highlighted.

## Screenshots

This is the initial view:

image

`contributors.txt` is staged:

image

`contributors.txt` is unstaged (either by double clicking on the file name or by pressing <kbd>U</kbd> while the file is selected:

image

The result of pressing <kbd>&darr;</kbd>:

image

## Did this work in previous version of GitExtensions

Version 2.51 also had this problem.

## Environment

- GitExtensions version: 3.00.00.4433
- GIT version: 2.20.1.windows.1
- OS version: Microsoft Windows NT 10.0.17134.0
- .NET version: .NET Framework 4.7.3324.0

## Proposed fix

https://github.com/monoblaine/gitextensions/commit/62e5287a8a2f4e8bb332bf9711ef8b0df25dfbfc#diff-16d3bf20df9b2330e5146f334a6e7f04

After some debugging I've found that the method `RestoreSelectedFiles` (in `FormCommit.cs`) does not call `SelectStoredNextIndex` after updating `_currentFilesList.SelectedItems`. But it *should* call that local function bacause updating `SelectedItems` makes a call to `StoreNextIndexToSelect` (in `FileStatusList.cs`)

*(This is my first contribution attempt. Hope to send a PR if everything's all right.)*

### LINKED ISSUE #7246: Commit dialog, FileStatusList: Up/down selects unexpected file
## Current behaviour

- pushing `Up`/`Down` keys does not have an effect or selects an unexpected file, e.g. after refresh

## Expected behaviour

- the file selection can correctly be moved using the keyboard 

## Steps to reproduce

- open the Commit dialog
- have at least two unstaged files
- select the last
- refresh (`F5`)
- hit `Up` => nothing happens
- hit `Down` => the first file is selected

## Did this work in previous version of GitExtensions

Have not checked

## Environment

- Git Extensions 3.3.0
- Build 00d5e2db95a84a5c3efa153fa38fb0a4a9397fbb
- Git 2.23.0.windows.1
- Microsoft Windows NT 10.0.18362.0
- .NET Framework 4.8.4010.0
- DPI 96dpi (no scaling)

### PR DESCRIPTION
**Pull request by mstv**
Fixes #7246 
Fixes #6212

## Proposed changes

- Set FileStatusListView.FocusedItem, too

## Test methodology 

- manual test (see STR in #7246)
- added NUnit test

## Test environment(s)

- Git Extensions 3.3.0
- Build 420bcdfe3fab4baef6a69b62698151b3170c2b35
- Git 2.23.0.windows.1
- Microsoft Windows NT 10.0.18362.0
- .NET Framework 4.8.4010.0
- DPI 96dpi (no scaling)

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

**Comment by mstv** [comment]
@mstfbot, merge in 1 day

**Comment by mstv** [comment]
@msftbot, merge in 1 day

**Comment by mstv** [comment]
@msftbot, merge in 16 hours

**Comment by RussKie** [comment]
I'm on a phone, but it looks like we have an inconsistent behaviour for `SelectedItem` vs `SelectedItems` as far as the focus concerned.
Could you please review both and ensure they are aligned in behaviour.

A set of tests would be good too.

**Comment by mstv** [comment]
You're so right, applies to `SelectedIndex`, too: Redundant interface implemented using Copy & Paste! 😠

**Comment by RussKie** [comment]
We are following CPD - copy-paste development :D

File: GitUI/UserControls/FileStatusList.cs [hunk 1]
```diff
@@ -592,25 +558,7 @@ int GetLastIndex()
...
-            {
-                SuspendLayout();
-
-                var itemCount = AllItems.Count();
-
-                for (var i = 0; i < itemCount; i++)
-                {
-                    FileStatusListView.Items[i].Selected = true;
-                    i++;
-                }
```
Comment (mstv) [inline GitUI/UserControls/FileStatusList.cs]: another bug fixed

File: GitUI/UserControls/FileStatusList.cs [hunk 2]
```diff
@@ -826,6 +755,38 @@ private async Task OpenSubmoduleAsync()
             process.Start();
         }
 
+        private void SelectItems(Func<ListViewItem, bool> isToBeSelectedFunc)
+        {
+            try
+            {
+                SuspendLayout();
```
Comment (mstv) [inline GitUI/UserControls/FileStatusList.cs]: I guess, this is pointless, isn't it? (factored out from some of the replaced functions)

**Comment by mstv** [comment]
WIP only because of missing NUnit tests

File: GitUI/UserControls/FileStatusList.cs [hunk 3]
```diff
@@ -727,28 +675,9 @@ public void SetSelectedIndex(int idx, bool notify)
...
-                    return FileStatusListView.SelectedIndices.Count;
-                }
-                finally
-                {
-                    ResumeLayout(true);
-                }
-            }
+            var regex = RegexForSelecting(selectionFilter);
+            SelectItems(item => regex.IsMatch(item.Name));
+            return FileStatusListView.SelectedIndices.Count;
```
Comment (mstv) [inline GitUI/UserControls/FileStatusList.cs]: Almost hidden feature: Have you ever hit `Ctrl+F` in the Unstage Files List?

Comment (RussKie) [reply GitUI/UserControls/FileStatusList.cs]: 👀 
No... That feature looks very strange.
We have a filter dropdown that is much more useful. The same filter should be in the Staged list too (and the Diff form too).

I would recommend that feature for removal.

File: GitUI/UserControls/FileStatusList.cs [hunk 4]
```diff
@@ -298,59 +298,47 @@ public int SelectedIndex
...
-                            newSelected = item;
-                            break;
+                            if (newSelected == null)
+                            {
+                                newSelected = item;
+                            }
+                            else if (gitItemStatus == status)
+                            {
+                                return item;
+                            }
```
Comment (RussKie) [inline GitUI/UserControls/FileStatusList.cs]: Can we please invert it to be:
```cs
if (status.CompareTo(gitItemStatus) == 0)
{
    if (gitItemStatus == status) {
        return item;
    }

    if (newSelected == null) { ... }
}
```

File: GitUI/UserControls/FileStatusList.cs [hunk 5]
```diff
@@ -826,6 +755,38 @@ private async Task OpenSubmoduleAsync()
             process.Start();
         }
 
+        private void SelectItems(Func<ListViewItem, bool> isToBeSelectedFunc)
```
[the remark this answers is not in the document]
Comment (mstv) [reply GitUI/UserControls/FileStatusList.cs]: > Linq has relatively simple arg names, like func, predicate, source etc.

Yes, I started with `predicate` and thought the readability of the imlementation could be better.
Switched back.

File: GitUI/UserControls/FileStatusList.cs (hunk 4 above)
Comment (mstv) [reply GitUI/UserControls/FileStatusList.cs]: This "inversion" will not have the same behavior. Note the `else` with respect to the `foreach` loop.

I have only moved this - not very self-explaining - logic. I don't want to change it.

Comment (RussKie) [reply GitUI/UserControls/FileStatusList.cs]: You are right, I was tired and didn't think it through.

**Pull request merged**